From 2a67666ad4677e3b11e7740d71afbf96a8a42541 Mon Sep 17 00:00:00 2001
From: Sergey Fedorov <barracuda@macos-powerpc.org>
Date: Thu, 20 Aug 2026 22:16:10 +0000
Subject: [PATCH 01/36] cmake: restore dual Qt4/Qt5 build system

Reintroduce WITH_QT5 (default ON) and branch package discovery on it,
mirroring the pre-af84b80c/30094392 Qt4 CMake support from v0.6: Qt4
find_package + trojita_check_qt4_module gate, per-target
qt5_use_modules()/target_link_libraries() pairs, qt4_wrap_ui/
qt4_add_resources fallbacks, and QtKeychain (Qt4 flavor) lookup.

QMimeDatabase is no longer vendored (30094392 dropped the bundled
src/mimetypes-qt4/ tree); the Qt4 branch instead locates the
standalone QtMimeTypes library (d1vanov/qt4-mimetypes, packaged as
devel/qt4-mimetypes) via pkg-config with a find_path/find_library
fallback, and links it into Composer and DesktopGui, the two library
targets whose sources #include <QMimeDatabase>. WITH_CRYPTO_MESSAGES
now also depends on WITH_QT5 since GpgME++'s Qt integration is
Qt5-only.

Restores trojita_check_qt4_module(), deleted by 25fa59a3.
---
 CMakeLists.txt            | 207 +++++++++++++++++++++++++++++++-------
 cmake/TrojitaOption.cmake |  26 +++++
 2 files changed, 195 insertions(+), 38 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 19fd9063..6389b431 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -55,6 +55,7 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CXX11_COMPILER_FLAGS}")
 include(TrojitaOption)
 
 trojita_option(WITH_DESKTOP "Build desktop version" ON)
+trojita_option(WITH_QT5 "Build with Qt5 library" ON)
 trojita_option(WITH_DBUS "Build with DBus library" AUTO)
 trojita_option(WITH_RAGEL "Build with Ragel library" AUTO)
 trojita_option(WITH_ZLIB "Build with zlib library" AUTO)
@@ -73,18 +74,57 @@ else()
     set(QTKEYCHAIN_DEPENDS "")
 endif()
 
-find_package(Qt5Core 5.2 REQUIRED)
-find_package(Qt5Gui REQUIRED)
-find_package(Qt5Network REQUIRED)
-find_package(Qt5Sql REQUIRED)
-find_package(Qt5WebKitWidgets REQUIRED)
-find_package(Qt5Widgets REQUIRED)
-find_package(Qt5LinguistTools)
-find_package(Qt5Svg REQUIRED)
-trojita_find_package(Qt5DBus "" "http://qt-project.org" "Qt5 D-Bus support" "Needed for IPC and for some plugins" WITH_DBUS)
-trojita_find_package(Qt5Test "" "http://qt-project.org" "Qt5 QTest library" "Needed for automated tests" WITH_TESTS)
-if(Qt5LinguistTools_FOUND)
-    find_package(Qt5LinguistForTrojita)
+if(WITH_QT5)
+    message(STATUS "Building the Qt5 version")
+    find_package(Qt5Core 5.2 REQUIRED)
+    find_package(Qt5Gui REQUIRED)
+    find_package(Qt5Network REQUIRED)
+    find_package(Qt5Sql REQUIRED)
+    find_package(Qt5WebKitWidgets REQUIRED)
+    find_package(Qt5Widgets REQUIRED)
+    find_package(Qt5LinguistTools)
+    find_package(Qt5Svg REQUIRED)
+    trojita_find_package(Qt5DBus "" "http://qt-project.org" "Qt5 D-Bus support" "Needed for IPC and for some plugins" WITH_DBUS)
+    trojita_find_package(Qt5Test "" "http://qt-project.org" "Qt5 QTest library" "Needed for automated tests" WITH_TESTS)
+    if(Qt5LinguistTools_FOUND)
+        find_package(Qt5LinguistForTrojita)
+    endif()
+else()
+    message(STATUS "Building the Qt4 version")
+    set(QT_USE_QTNETWORK 1)
+    set(QT_USE_QTSQL 1)
+    set(QT_USE_QTWEBKIT 1)
+    set(QT_USE_QTSVG 1)
+    if(WITH_DBUS)
+        set(QT_USE_QTDBUS 1)
+    endif()
+    if(WITH_TESTS)
+        set(QT_USE_QTTEST 1)
+    endif()
+    trojita_find_package(Qt4 4.8 "http://qt-project.org" "Qt4" "Needed for building" REQUIRED)
+    include(${QT_USE_FILE})
+    trojita_check_qt4_module(QTCORE REQUIRED)
+    trojita_check_qt4_module(QTGUI REQUIRED)
+    trojita_check_qt4_module(QTNETWORK REQUIRED)
+    trojita_check_qt4_module(QTSQL REQUIRED)
+    trojita_check_qt4_module(QTWEBKIT REQUIRED)
+    trojita_check_qt4_module(QTSVG REQUIRED)
+    trojita_check_qt4_module(QTDBUS WITH_DBUS)
+    trojita_check_qt4_module(QTTEST WITH_TESTS)
+
+    find_package(PkgConfig)
+    if(PKG_CONFIG_FOUND)
+        pkg_check_modules(QTMIMETYPES QtMimeTypes)
+    endif()
+    if(NOT QTMIMETYPES_FOUND)
+        find_path(QTMIMETYPES_INCLUDE_DIRS NAMES qmimedatabase.h PATH_SUFFIXES QtMimeTypes)
+        find_library(QTMIMETYPES_LIBRARIES NAMES QtMimeTypes)
+    endif()
+    if(NOT QTMIMETYPES_INCLUDE_DIRS OR NOT QTMIMETYPES_LIBRARIES)
+        message(FATAL_ERROR "QtMimeTypes (port devel/qt4-mimetypes, https://github.com/d1vanov/qt4-mimetypes) is required for the Qt4 build")
+    endif()
+    message(STATUS "Found QtMimeTypes (includes at ${QTMIMETYPES_INCLUDE_DIRS}, lib at ${QTMIMETYPES_LIBRARIES})")
+    include_directories(${QTMIMETYPES_INCLUDE_DIRS})
 endif()
 
 trojita_plugin_option(WITH_ABOOKADDRESSBOOK_PLUGIN "Build AbookAddressbook plugin" STATIC)
@@ -95,7 +135,7 @@ trojita_find_package(Git "" "" "" "")
 
 trojita_find_package(Mimetic "" "http://www.codesink.org/mimetic_mime_library.html" "C++ MIME Library" "Required for client-side MIME parsing" WITH_MIMETIC)
 trojita_find_package(KF5Gpgmepp "" "https://quickgit.kde.org/?p=gpgmepp.git" "C++ bindings for gpgme" "Needed for encrypted/signed e-mails" WITH_GPGMEPP)
-trojita_option(WITH_CRYPTO_MESSAGES "Enable support for encrypted messages" AUTO "WITH_MIMETIC;WITH_GPGMEPP")
+trojita_option(WITH_CRYPTO_MESSAGES "Enable support for encrypted messages" AUTO "WITH_MIMETIC;WITH_GPGMEPP;WITH_QT5")
 
 if(WIN32)
     trojita_find_package(MakeNSIS "" "http://nsis.sourceforge.net" "Nullsoft Scriptable Install System" "Needed for building Windows installer" WITH_NSIS)
@@ -112,7 +152,11 @@ if(WIN32)
     endif()
 endif()
 
-trojita_find_package(Qt5Keychain QUIET "https://github.com/frankosterfeld/qtkeychain" "QtKeychain library (Qt5 version)" "Needed for QtKeychain password plugin" WITH_QTKEYCHAIN_PLUGIN)
+if(WITH_QT5)
+    trojita_find_package(Qt5Keychain QUIET "https://github.com/frankosterfeld/qtkeychain" "QtKeychain library (Qt5 version)" "Needed for QtKeychain password plugin" WITH_QTKEYCHAIN_PLUGIN)
+else()
+    trojita_find_package(QtKeychain QUIET "https://github.com/frankosterfeld/qtkeychain" "QtKeychain library" "Needed for QtKeychain password plugin" WITH_QTKEYCHAIN_PLUGIN)
+endif()
 if(Qt5Keychain_FOUND OR QtKeychain_FOUND)
     message(STATUS "Found QtKeychain library (includes at ${QTKEYCHAIN_INCLUDE_DIRS}, lib at ${QTKEYCHAIN_LIBRARIES})")
 else()
@@ -558,13 +602,21 @@ set_source_files_properties(${version_files}
 add_library(Common STATIC ${libCommon_SOURCES})
 set_property(TARGET Common APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
 add_dependencies(Common version)
-qt5_use_modules(Common Core Network)
+if(WITH_QT5)
+    qt5_use_modules(Common Core Network)
+else()
+    target_link_libraries(Common ${QT_QTCORE_LIBRARY} ${QT_QTNETWORK_LIBRARY})
+endif()
 
 add_library(AppVersion STATIC ${libAppVersion_SOURCES})
 set_property(TARGET AppVersion APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
 add_dependencies(AppVersion version)
 target_link_libraries(AppVersion Common)
-qt5_use_modules(AppVersion Core)
+if(WITH_QT5)
+    qt5_use_modules(AppVersion Core)
+else()
+    target_link_libraries(AppVersion ${QT_QTCORE_LIBRARY})
+endif()
 
 if(WITH_SHARED_PLUGINS)
     add_library(Plugins SHARED ${libPlugins_SOURCES})
@@ -575,12 +627,20 @@ else()
 endif()
 set_target_properties(Plugins PROPERTIES OUTPUT_NAME trojita_plugins)
 set_property(TARGET Plugins APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
-qt5_use_modules(Plugins Core)
+if(WITH_QT5)
+    qt5_use_modules(Plugins Core)
+else()
+    target_link_libraries(Plugins ${QT_QTCORE_LIBRARY})
+endif()
 
 add_library(UiUtils STATIC ${libUiUtils_SOURCES})
 set_property(TARGET UiUtils APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
 target_link_libraries(UiUtils Plugins Common)
-qt5_use_modules(UiUtils Core Gui Network)
+if(WITH_QT5)
+    qt5_use_modules(UiUtils Core Gui Network)
+else()
+    target_link_libraries(UiUtils ${QT_QTCORE_LIBRARY} ${QT_QTGUI_LIBRARY} ${QT_QTNETWORK_LIBRARY})
+endif()
 
 add_library(Streams STATIC ${libStreams_SOURCES})
 set_property(TARGET Streams APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
@@ -588,36 +648,68 @@ if(WITH_ZLIB)
     set_property(TARGET Streams APPEND PROPERTY INCLUDE_DIRECTORIES ${ZLIB_INCLUDE_DIR})
     target_link_libraries(Streams ${ZLIB_LIBRARIES})
 endif()
-qt5_use_modules(Streams Network)
+if(WITH_QT5)
+    qt5_use_modules(Streams Network)
+else()
+    target_link_libraries(Streams ${QT_QTCORE_LIBRARY} ${QT_QTNETWORK_LIBRARY})
+endif()
 
 add_library(IPC STATIC ${libIPC_SOURCES})
 set_property(TARGET IPC APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
 if(WITH_DBUS)
-    qt5_use_modules(IPC DBus Widgets)
+    if(WITH_QT5)
+        qt5_use_modules(IPC DBus Widgets)
+    else()
+        target_link_libraries(IPC ${QT_QTCORE_LIBRARY} ${QT_QTGUI_LIBRARY} ${QT_QTDBUS_LIBRARY})
+    endif()
 else()
-    qt5_use_modules(IPC Core)
+    if(WITH_QT5)
+        qt5_use_modules(IPC Core)
+    else()
+        target_link_libraries(IPC ${QT_QTCORE_LIBRARY})
+    endif()
 endif()
 
 add_library(qwwsmtpclient STATIC ${libqwwsmtpclient_SOURCES})
-qt5_use_modules(qwwsmtpclient Network)
+if(WITH_QT5)
+    qt5_use_modules(qwwsmtpclient Network)
+else()
+    target_link_libraries(qwwsmtpclient ${QT_QTCORE_LIBRARY} ${QT_QTNETWORK_LIBRARY})
+endif()
 
 add_library(MSA STATIC ${libMSA_SOURCES})
 set_property(TARGET MSA APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
 target_link_libraries(MSA Imap Streams qwwsmtpclient)
-qt5_use_modules(MSA Network)
+if(WITH_QT5)
+    qt5_use_modules(MSA Network)
+else()
+    target_link_libraries(MSA ${QT_QTCORE_LIBRARY} ${QT_QTNETWORK_LIBRARY})
+endif()
 
 add_library(Composer STATIC ${libComposer_SOURCES})
 set_property(TARGET Composer APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
 target_link_libraries(Composer Common MSA Streams UiUtils qwwsmtpclient)
-qt5_use_modules(Composer Gui Network)
+if(WITH_QT5)
+    qt5_use_modules(Composer Gui Network)
+else()
+    target_link_libraries(Composer ${QT_QTCORE_LIBRARY} ${QT_QTGUI_LIBRARY} ${QT_QTNETWORK_LIBRARY} ${QTMIMETYPES_LIBRARIES})
+endif()
 
 add_library(Imap STATIC ${libImap_SOURCES})
 set_property(TARGET Imap APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
 target_link_libraries(Imap Common Streams UiUtils)
-qt5_use_modules(Imap Gui Network Sql)
+if(WITH_QT5)
+    qt5_use_modules(Imap Gui Network Sql)
+else()
+    target_link_libraries(Imap ${QT_QTCORE_LIBRARY} ${QT_QTGUI_LIBRARY} ${QT_QTNETWORK_LIBRARY} ${QT_QTSQL_LIBRARY})
+endif()
 
 add_library(Cryptography STATIC ${libCryptography_SOURCES})
-qt5_use_modules(Cryptography Core)
+if(WITH_QT5)
+    qt5_use_modules(Cryptography Core)
+else()
+    target_link_libraries(Cryptography ${QT_QTCORE_LIBRARY})
+endif()
 set_property(TARGET Cryptography APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
 target_link_libraries(Cryptography Common Imap)
 if(WITH_MIMETIC)
@@ -642,7 +734,11 @@ endif()
 ## QtKeyChain plugin
 if(WITH_QTKEYCHAIN_PLUGIN)
     trojita_add_plugin(trojita_plugin_QtKeychainPasswordPlugin WITH_QTKEYCHAIN_PLUGIN src/Plugins/QtKeyChain/QtKeyChainPassword.cpp)
-    qt5_use_modules(trojita_plugin_QtKeychainPasswordPlugin Core DBus)
+    if(WITH_QT5)
+        qt5_use_modules(trojita_plugin_QtKeychainPasswordPlugin Core DBus)
+    else()
+        target_link_libraries(trojita_plugin_QtKeychainPasswordPlugin ${QT_QTCORE_LIBRARY} ${QT_QTDBUS_LIBRARY})
+    endif()
     target_link_libraries(trojita_plugin_QtKeychainPasswordPlugin ${QTKEYCHAIN_LIBRARIES})
     set_property(TARGET trojita_plugin_QtKeychainPasswordPlugin APPEND PROPERTY INCLUDE_DIRECTORIES ${QTKEYCHAIN_INCLUDE_DIRS})
 endif()
@@ -663,12 +759,20 @@ if(WITH_ABOOKADDRESSBOOK_PLUGIN)
         ${path_AbookAddressbook}/onecontact.ui
     )
 
-    qt5_wrap_ui(libAbookAddressbook_UI_OUT ${libAbookAddressbook_UI})
+    if(WITH_QT5)
+        qt5_wrap_ui(libAbookAddressbook_UI_OUT ${libAbookAddressbook_UI})
+    else()
+        qt4_wrap_ui(libAbookAddressbook_UI_OUT ${libAbookAddressbook_UI})
+    endif()
 
     trojita_add_plugin(trojita_plugin_AbookAddressbookPlugin WITH_ABOOKADDRESSBOOK_PLUGIN ${libAbookAddressbook_SOURCES} ${libAbookAddressbook_UI_OUT})
     set_property(TARGET trojita_plugin_AbookAddressbookPlugin APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
 
-    qt5_use_modules(trojita_plugin_AbookAddressbookPlugin Widgets)
+    if(WITH_QT5)
+        qt5_use_modules(trojita_plugin_AbookAddressbookPlugin Widgets)
+    else()
+        target_link_libraries(trojita_plugin_AbookAddressbookPlugin ${QT_QTGUI_LIBRARY})
+    endif()
 
     set(be_contacts_SOURCES
         ${path_AbookAddressbook}/main.cpp
@@ -681,7 +785,11 @@ if(WITH_ABOOKADDRESSBOOK_PLUGIN)
         set_property(TARGET be.contacts APPEND PROPERTY COMPILE_DEFINITIONS QT_STATICPLUGIN)
         target_link_libraries(be.contacts trojita_plugin_AbookAddressbookPlugin)
     endif()
-    qt5_use_modules(be.contacts Widgets)
+    if(WITH_QT5)
+        qt5_use_modules(be.contacts Widgets)
+    else()
+        target_link_libraries(be.contacts ${QT_QTGUI_LIBRARY} ${QT_QTCORE_LIBRARY})
+    endif()
 endif()
 
 # Generate file static_plugins.h.in
@@ -693,8 +801,13 @@ endforeach()
 execute_process(COMMAND ${CMAKE_COMMAND} -E copy_if_different ${CMAKE_CURRENT_BINARY_DIR}/static_plugins.h.in ${CMAKE_CURRENT_BINARY_DIR}/static_plugins.h)
 
 if(WITH_DESKTOP)
-    qt5_wrap_ui(libDesktopGui_UI_OUT ${libDesktopGui_UI})
-    qt5_add_resources(libDesktopGui_RESOURCES_OUT ${libDesktopGui_RESOURCES})
+    if(WITH_QT5)
+        qt5_wrap_ui(libDesktopGui_UI_OUT ${libDesktopGui_UI})
+        qt5_add_resources(libDesktopGui_RESOURCES_OUT ${libDesktopGui_RESOURCES})
+    else()
+        qt4_wrap_ui(libDesktopGui_UI_OUT ${libDesktopGui_UI})
+        qt4_add_resources(libDesktopGui_RESOURCES_OUT ${libDesktopGui_RESOURCES})
+    endif()
 
     add_library(DesktopGui STATIC ${libDesktopGui_SOURCES} ${libDesktopGui_UI_OUT} ${libDesktopGui_RESOURCES_OUT})
     set_property(TARGET DesktopGui APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
@@ -708,8 +821,13 @@ if(WITH_DESKTOP)
     add_executable(trojita WIN32 ${trojita_desktop_SOURCES} ${trojita_QM})
     set_property(TARGET trojita APPEND PROPERTY COMPILE_DEFINITIONS QT_NO_CAST_FROM_ASCII QT_NO_CAST_TO_ASCII)
     target_link_libraries(trojita AppVersion Common UiUtils DesktopGui ${STATIC_PLUGINS})
-    qt5_use_modules(DesktopGui Network WebKitWidgets)
-    qt5_use_modules(trojita Widgets Network)
+    if(WITH_QT5)
+        qt5_use_modules(DesktopGui Network WebKitWidgets)
+        qt5_use_modules(trojita Widgets Network)
+    else()
+        target_link_libraries(DesktopGui ${QT_QTCORE_LIBRARY} ${QT_QTGUI_LIBRARY} ${QT_QTNETWORK_LIBRARY} ${QT_QTWEBKIT_LIBRARY} ${QTMIMETYPES_LIBRARIES})
+        target_link_libraries(trojita ${QT_QTGUI_LIBRARY} ${QT_QTCORE_LIBRARY})
+    endif()
 endif()
 
 
@@ -743,7 +861,11 @@ if(WITH_TESTS)
         tests/Utils/LibMailboxSync.cpp
     )
     add_library(test_LibMailboxSync STATIC ${test_LibMailboxSync_SOURCES})
-    qt5_use_modules(test_LibMailboxSync Test Network)
+    if(WITH_QT5)
+        qt5_use_modules(test_LibMailboxSync Test Network)
+    else()
+        target_link_libraries(test_LibMailboxSync ${QT_QTCORE_LIBRARY} ${QT_QTNETWORK_LIBRARY} ${QT_QTTEST_LIBRARY})
+    endif()
     set_property(TARGET test_LibMailboxSync APPEND PROPERTY INCLUDE_DIRECTORIES
         ${CMAKE_CURRENT_SOURCE_DIR}/tests
         ${CMAKE_CURRENT_SOURCE_DIR}/tests/Utils)
@@ -753,7 +875,11 @@ if(WITH_TESTS)
         set(test_${fname}_SOURCES tests/${dir}/test_${fname}.cpp)
         add_executable(test_${fname} ${test_${fname}_SOURCES})
         target_link_libraries(test_${fname} Imap MSA Streams Common Composer Cryptography test_LibMailboxSync)
-        qt5_use_modules(test_${fname} Network Sql Test Widgets)
+        if(WITH_QT5)
+            qt5_use_modules(test_${fname} Network Sql Test Widgets)
+        else()
+            target_link_libraries(test_${fname} ${QT_QTCORE_LIBRARY} ${QT_QTGUI_LIBRARY} ${QT_QTNETWORK_LIBRARY} ${QT_QTSQL_LIBRARY} ${QT_QTTEST_LIBRARY})
+        endif()
         set_property(TARGET test_${fname} APPEND PROPERTY INCLUDE_DIRECTORIES ${CMAKE_CURRENT_SOURCE_DIR}/tests)
         if(NOT CMAKE_CROSSCOMPILING)
             add_test(test_${fname} test_${fname})
@@ -764,8 +890,13 @@ if(WITH_TESTS)
     trojita_test(Composer Composer_Submission)
     trojita_test(Composer Composer_responses)
     trojita_test(Composer Html_formatting)
-    qt5_use_modules(test_Composer_responses WebKitWidgets)
-    qt5_use_modules(test_Html_formatting WebKitWidgets)
+    if(WITH_QT5)
+        qt5_use_modules(test_Composer_responses WebKitWidgets)
+        qt5_use_modules(test_Html_formatting WebKitWidgets)
+    else()
+        target_link_libraries(test_Composer_responses ${QT_QTWEBKIT_LIBRARY})
+        target_link_libraries(test_Html_formatting ${QT_QTWEBKIT_LIBRARY})
+    endif()
     trojita_test(Imap Imap_DisappearingMailboxes)
     trojita_test(Imap Imap_Idle)
     trojita_test(Imap Imap_LowLevelParser)
diff --git a/cmake/TrojitaOption.cmake b/cmake/TrojitaOption.cmake
index ed99e952..2e7b8ec6 100644
--- a/cmake/TrojitaOption.cmake
+++ b/cmake/TrojitaOption.cmake
@@ -106,6 +106,32 @@ macro(trojita_add_plugin target type)
     endif()
 endmacro()
 
+# trojita_check_qt4_module(module [options...])
+macro(trojita_check_qt4_module module)
+    if(NOT QT_${module}_FOUND)
+        foreach(arg ${ARGN})
+            if("${${arg}}" STREQUAL "ON" OR "${arg}" STREQUAL REQUIRED)
+                message(STATUS "Qt4 module ${module} is required because of ${arg}")
+                message(FATAL_ERROR "Qt4 module ${module} was not found")
+            elseif("${${arg}}" STREQUAL "AUTO")
+                message(STATUS "Disabling ${arg} because Qt4 module ${module} was not found")
+                if(NOT ${arg} MATCHES "^${arg}$")
+                    set(${arg} "${${arg}}" CACHE INTERNAL "")
+                endif()
+                set(${arg} OFF)
+                get_property(features GLOBAL PROPERTY ENABLED_FEATURES)
+                list(REMOVE_ITEM features ${arg})
+                set_property(GLOBAL PROPERTY ENABLED_FEATURES "${features}")
+                get_property(features GLOBAL PROPERTY DISABLED_FEATURES)
+                list(FIND features ${arg} id)
+                if(id EQUAL -1)
+                    set_property(GLOBAL APPEND PROPERTY DISABLED_FEATURES "${arg}")
+                endif()
+            endif()
+        endforeach()
+    endif()
+endmacro()
+
 # trojita_find_package(package version url description purpose [options...])
 macro(trojita_find_package package version url description purpose)
     set(type OPTIONAL)
-- 
2.43.0

